odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+        ----------
+        x1, x2 : `CupyTensor`
+            Dividend and divisor in the quotient.
+        out : `CupyTensor`
+            Tensor to which the result is written.
+
+        Examples
+        --------
+        Out-of-place evaluation:
+
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> y = r3.element([-1, 2, 1])
+        >>> r3.divide(x, y)
+        rn(3, impl='cupy').element([-1.,  1.,  3.])
+
+        In-place:
+
+        >>> out = r3.element()
+        >>> result = r3.divide(x, y, out=out)
+        >>> result
+        rn(3, impl='cupy').element([-1.,  1.,  3.])
+        >>> result is out
+        True
+        """
+        x1.ufuncs.divide(x2, out=out)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        if self.ndim == 1:
+            posargs = [self.size]
+        else:
+            posargs = [self.shape]
+
+        if self.is_real:
+            constructor_name = 'rn'
+        elif self.is_complex:
+            constructor_name = 'cn'
+        else:
+            constructor_name = 'tensor_space'
+
+        if (constructor_name == 'tensor_space' or
+                (not self.is_real and not self.is_complex) or
+                self.dtype != self.default_dtype(self.field)):
+            posargs.append(dtype_str(self.dtype))
+
+        optargs = [('impl', self.impl, 'numpy'),  # for the helper functions
+                   ('device', self.device, cupy.cuda.get_device_id())]
+        inner_str = signature_string(posargs, optargs)
+        weight_str = self.weighting.repr_part
+        if weight_str:
+            inner_str += ', ' + weight_str
+
+        return '{}({})'.format(constructor_name, inner_str)
+
+    @property
+    def element_type(self):
+        """`CupyTensor`"""
+        return CupyTensor
+
+    @staticmethod
+    def available_dtypes():
+        """Return the data types available for this space."""
+        dtypes = (np.sctypes['float'] +
+                  np.sctypes['complex'] +
+                  np.sctypes['int']
+                  + np.sctypes['uint']
+                  + [bool])
+        dtypes.remove(np.float128)
+        dtypes.remove(np.complex256)
+        return tuple(np.dtype(dtype) for dtype in dtypes)
+
+    @staticmethod
+    def default_dtype(field=None):
+        """Return the default data type of this space type for a given field.
+
+        Parameters
+        ----------
+        field : `Field`, optional
+            Set of numbers to be represented by a data type.
+            Currently supported : `RealNumbers`, `ComplexNumbers`.
+            Default: `RealNumbers`
+
+        Returns
+        -------
+        dtype : `numpy.dtype`
+            Numpy data type specifier. The returned defaults are:
+
+            ``RealNumbers()`` : ``np.dtype('float64')``
+
+            ``ComplexNumbers()`` : not supported
+        """
+        if field is None or field == RealNumbers():
+            return np.dtype('float64')
```

Comment:
Should be float32 on GPU systems for sure.

Reply:
Shouldn't we go with the default of the backend?

Reply:
Agree, I just expected it to be float32

Reply:
Well the question was open. If we find single precision more reasonable we can go for that as well.

Reply:
Well what does the backend use? This http://docs.chainer.org/en/v1.21.0/_modules/cupy/creation/basic.html tells it it's float64, and then we should follow that I guess.

Reply:
Okay. They try to stick as closely to Numpy, which is good for us. I'll go with their default then.

Reply:
Please add some comment on this in this function (i.e. "we use the same as...") for future readers

Reply:
Done.